artsy/force
# [Refactor] Remove bucket-assets, replace with Webpack hashing

Replaces `bucket-assets` functionality with Webpack and other local files. 

Overview:

1) Replaced un-cacheable bucket-asset fingerprints with consistent hashes from webpack
1) This gives us source maps! 
1) Replaced bucket-assets upload functionality with a local script, `scripts/uploadToS3.js`
1) Added a script to compile / hash CSS, which will change per build (like it did before). This doesn't go through webpack because our code is too old and it would be hell to fix; its something we'll have to live with for now and is increasingly irrelevant as we move more apps over to styled-components. The hashes get merged into webpacks manifest as a seed
1) Added a new middleware helper to return manifest assets; only applies to prod
1) Removed `assets.sh` and `publish_assets.sh`, instead calling from package scripts
1) Added [Brotli](https://en.wikipedia.org/wiki/Brotli) compression support, which reduces our js bundle sizes substantially in supported browsers. 

To test, pull down branch and run 
```
yarn publish-assets && yarn start:prod
```
This will upload assets to a temporary S3 bucket and then boot force with the manifest. (Note that this assumes you have S3 credentials in your `.env`.)

The last remaining thing is to pull out all of the docker fingerprinting stuff since it's no longer needed, thank the lord 🎉Rolling back and forward between containers is enough. 

TODO: 
- [x] Define flag: `ENABLE_EXPERIMENTAL_ASSET_BUNDLING`
- [x] Add Webpack hashing pipeline
- [x] Add css hashing pipeline 
- [x] Implement `express-static-gzip` for brotli / gzip compression
- [x] Update templates to point to hashed assets 
- [x] Upload assets to S3
- [x] Remove `ASSET_MANIFEST` hashing in docker, and other references
- [x] Remove temporary `artsy-chris-test-bucket` S3 bucket
- [x] Enable Brotli compression on CloudFront / investigate whats up with no gzip from S3
